// Copyright 2017 Intel Corporation
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
// -----------------------------------------------------------------------------

syntax = "proto3";

import "property.proto";


message Record {
  message AssociatedAgent {
    string agent_id = 1;
    uint64 timestamp = 2;
  }

  // The user-defined natural key which identifies the object in the
  // real world (for example a serial number)
  string record_id = 1;

  string record_type = 2;

  // Ordered oldest to newest by timestamp
  repeated AssociatedAgent owners = 3;
  repeated AssociatedAgent custodians = 4;

  // Flag indicating whether the Record can be updated. If it is set
  // to true, then the record has been finalized and no further
  // changes can be made to it or its Properties.
  bool final = 5;
}


message RecordContainer {
  repeated Record entries = 1;
}


message RecordType {
  // A unique human-readable designation for the RecordType
  string name = 1;

  repeated PropertySchema properties = 2;
}


message RecordTypeContainer {
  repeated RecordType entries = 1;
}
